### DISCUSSION TITLE
How to render ContentItem with other layout to string

### DISCUSSION THREAD

**Discussion by bashuss** [OrchardCMS/OrchardCore#9989] [Q&A] [answered]
Hi,

I tried to write a controller that takes a contentItemId and renders this contentItem to html. All the normal shapes should be used, but a different layout around it.

My approach was this:

**In the controller:**

```
var html = await liquidTemplateManager.RenderStringAsync(@$"{{{{ ""MyLayout"" | shape_new: Id:""{contentItemId}"" | shape_render }}}}""", _htmlEncoder);
```

**In Views Folder:** `MyLayout.liquid`
```

{% assign contentItem = Content.ContentItemId[Model.Id] %}
{% assign name = "Content__" | append: contentItem.ContentType %}
{% assign shape = name | shape_new: ContentItem:contentItem %}

    <!-- none of these work properly! First three do nothing --> 
    {{ contentItem | shape_render }}
    {{ contentItem.ContentItem | shape_render }}
    {{ contentItem.Content | shape_render }}
    <!-- this one renders the shape template with the name created above, but without contained content, e.g. HtmlBodyPart... -->
    {{ shape | shape_render }}
```

**How is this supposed to be done?**
`{{ contentItem.Content }}` works properly: it shows the contentItem as JSON
 `{{ contentItem.DisplayText }}, {{ contentItem.ContentType }}, {{ contentItem.ContentType }}` all work properly: the show the expected value as shown in documentation here https://docs.orchardcore.net/en/latest/docs/reference/modules/Liquid/#modelcontentitem

1. Where is the problem? Is it in the controller or in the liquid template?
2. I have found nothing in the documentation about how to get a contentItem by ContentItemId and render it's shape. I guess this should be mentioned somewhere?

Thanks a lot  for your insights,
Bernd

**Comment by Skrypt** [comment]
For #2 : https://docs.orchardcore.net/en/latest/docs/reference/modules/Contents/#loading-from-a-content-item-id

**Reply by bashuss** [reply]
Yes exactly - but there is not enough info: 
I do what is written there. It is the first line of my liquid template, as given above. But it is not possible to render this contentItem properly and the documentation does not show how to do that in this context.

So I do not know, if this is a bug or if I am doing something wrong. That is why I give the complete context in shortest possible form.

**Comment by ns8482e** [comment, marked as answer]
Try

```liquid
{% assign contentItem = Content.ContentItemId[Model.Id] %}
 {{ contentItem | shape_build_display: "Detail" | shape_render }}
```

**Reply by bashuss** [reply]
This does the job! Thanks!
So the idea with the controller was correct so far.

**Comment by bashuss** [comment]
One thing still does not work:
When putting a zone definition in my MyLayout.liquid template like this
```
 {% render_section "Header" %}
```
nothing is inserted there, even though the shape template of my contentItem rendered in MyLayout.liquid by

```
{{ contentItem | shape_build_display: "Detail" | shape_render }}
```
contains code that writes to the zone "Header" like this:
```
{% zone "Header" %}
    <h1>{{ Model.ContentItem | display_text }}</h1>
{% endzone %}
```

What am I missing here?

**Reply by Skrypt** [reply]
https://docs.orchardcore.net/en/latest/docs/reference/modules/Liquid/#layout

Are you assigning this custom MyLayout.liquid file?

**Reply by bashuss** [reply]
No, I am not, as I don't know exactly how to make that work with my MVC controller.
The documentation says, that `{% layout "MyLayout" %}` would be used to tell a View about the layout file, but I am not rendering inside a View...

My first idea would be something like this inside the Controller, but I did not test it yet... but it "feels" wrong somehow.
```
var html = await liquidTemplateManager.RenderStringAsync(
@$"{{% layout ""MyLayout"" %}}
{{% assign contentItem = Content.ContentItemId[""{contentItemId}""] %}}
{{{{ contentItem | shape_build_display: ""Detail"" | shape_render }}}}", _htmlEncoder);
```
